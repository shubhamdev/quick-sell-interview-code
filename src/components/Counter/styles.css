.counter {
  width: 180px;
  height: 56px;
  display: flex;
  flex-flow: row nowrap;
  justify-content: space-between;
  align-items: center;
  background-color: var(--secondary-color);
  border-radius: 8px;
  border: 1px solid var(--primary-color);
  overflow: hidden;
  box-sizing: border-box;
  color: var(--primary-color);
}

.count-input {
  width: 60px;
  height: 56px;
  background: none;
  text-align: center;
  color: var(--primary-color);
  margin: 0;
  padding: 0;
  border: none;
  border-radius: 0;
  outline: none;
  -moz-appearance: textfield;
}

.count-input::-webkit-outer-spin-button,
.count-input::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}