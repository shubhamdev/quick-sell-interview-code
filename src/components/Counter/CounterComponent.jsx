import { number, func } from "prop-types";

import Button from "../Button";
import "./styles.css";

const CounterComponent = ({
  value = 1,
  max = 1000,
  handleDecrement = () => {},
  handleIncrement = () => {},
  onChange = () => {},
}) => (
  <div className="counter">
    <Button
      label="-"
      width={60}
      height={56}
      color="primary"
      backgroundColor="tertiary"
      borderWidth="0 1px 0 0"
      borderColor="primary"
      onClick={handleDecrement}
    />
    <input
      type="number"
      className="count-input"
      aria-label="Count"
      value={value}
      max={max}
      onChange={onChange}
    />
    <Button
      label="+"
      width={60}
      height={56}
      color="tertiary"
      backgroundColor="primary"
      borderWidth="0 0 0 1px"
      borderColor="primary"
      disabled={value === max}
      onClick={handleIncrement}
    />
  </div>
);

CounterComponent.propTypes = {
  value: number,
  max: number,
  handleDecrement: func,
  handleIncrement: func,
  onChange: func,
};

export default CounterComponent;
