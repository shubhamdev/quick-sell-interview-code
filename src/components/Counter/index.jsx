import { number } from "prop-types";

import CounterComponent from "./CounterComponent";

const Counter = ({
  count = 1,
  max = 1000,
  handleDecrement = () => {},
  handleIncrement = () => {},
  onChange = () => {},
}) => {
  return (
    <CounterComponent
      value={count}
      max={max}
      handleDecrement={handleDecrement}
      handleIncrement={handleIncrement}
      onChange={onChange}
    />
  );
};

Counter.propTypes = {
  initialValue: number,
  max: number,
};

export default Counter;
