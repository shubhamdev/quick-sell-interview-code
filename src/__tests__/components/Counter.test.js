import { render, screen, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import pretty from "pretty";

import Counter from "../../components/Counter";
import CounterComponent from "../../components/Counter/CounterComponent";

describe("Test Counter Component", () => {
  const handleIncrement = jest.fn();
  const handleDecrement = jest.fn();
  const onChange = jest.fn();

  it("renders default Counter component", () => {
    const { container } = render(<Counter />);
    const counterComponent = document.querySelector(".counter");
    expect(counterComponent).toBeInTheDocument();
    expect(counterComponent).toHaveClass("counter");

    const decrementButton = screen.getByText(/-/i);
    expect(decrementButton).toBeInTheDocument();
    userEvent.click(decrementButton);
    expect(handleDecrement).toHaveBeenCalledTimes(0);

    const incrementButton = screen.getByText(/\+/i);
    expect(incrementButton).toBeInTheDocument();
    userEvent.click(incrementButton);
    expect(handleIncrement).toHaveBeenCalledTimes(0);

    const inputElement = screen.getByLabelText(/Count/gi);
    expect(inputElement).toBeInTheDocument();
    fireEvent.change(inputElement, {
      target: { value: "100" },
    });
    expect(onChange).toHaveBeenCalledTimes(0);

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it("renders Counter component with handlers", () => {
    const { container } = render(
      <CounterComponent
        handleDecrement={handleDecrement}
        handleIncrement={handleIncrement}
        onChange={onChange}
      />
    );
    const counterComponent = document.querySelector(".counter");
    expect(counterComponent).toBeInTheDocument();
    expect(counterComponent).toHaveClass("counter");

    const decrementButton = screen.getByText(/-/i);
    expect(decrementButton).toBeInTheDocument();
    userEvent.click(decrementButton);
    expect(handleDecrement).toHaveBeenCalledTimes(1);

    const incrementButton = screen.getByText(/\+/i);
    expect(incrementButton).toBeInTheDocument();
    userEvent.click(incrementButton);
    expect(handleIncrement).toHaveBeenCalledTimes(1);

    const inputElement = screen.getByLabelText(/Count/gi);
    expect(inputElement).toBeInTheDocument();
    fireEvent.change(inputElement, {
      target: { value: "100" },
    });
    expect(onChange).toHaveBeenCalledTimes(1);

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });
});
