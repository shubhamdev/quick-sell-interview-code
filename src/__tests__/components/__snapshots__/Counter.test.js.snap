// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Counter Component renders Counter component with handlers 1`] = `"<div class=\\"counter\\"><button style=\\"width: 60px; height: 56px; cursor: pointer; border-width: 0px 1px 0px 0px; border-style: solid;\\">-</button><input type=\\"number\\" class=\\"count-input\\" aria-label=\\"Count\\" max=\\"1000\\" value=\\"1\\"><button style=\\"width: 60px; height: 56px; cursor: pointer; border-width: 0px 0px 0px 1px; border-style: solid;\\">+</button></div>"`;

exports[`Test Counter Component renders default Counter component 1`] = `"<div class=\\"counter\\"><button style=\\"width: 60px; height: 56px; cursor: pointer; border-width: 0px 1px 0px 0px; border-style: solid;\\">-</button><input type=\\"number\\" class=\\"count-input\\" aria-label=\\"Count\\" max=\\"1000\\" value=\\"100\\"><button style=\\"width: 60px; height: 56px; cursor: pointer; border-width: 0px 0px 0px 1px; border-style: solid;\\">+</button></div>"`;
